{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0b3c7140-b35b-48c1-84a1-c480fc5b67b9",
   "metadata": {},
   "source": [
    "# Annotation of GENIE mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "984c9f3d-781f-41a7-b043-360c4765a5d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import nirvanaparser as nip\n",
    "from tqdm.auto import tqdm\n",
    "import gzip"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ac23242-fa8e-4a57-9d3b-c5690a7e8da1",
   "metadata": {},
   "source": [
    "## Input files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9a624239-6d96-4693-b52f-cceb4b9a602e",
   "metadata": {},
   "outputs": [],
   "source": [
    "genie_dir = \"../Data/Original/genie-15.0\"\n",
    "vcf_file = os.path.join(genie_dir, \"genie_normalized.vcf.bgz\")\n",
    "json_file = \"../Results/genie_normalized.json.gz\"\n",
    "cosmic_cmc_file = os.path.join(\"../Data/Original\", \"cmc_export.tsv.gz\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d01fa737-85e4-4769-b6d9-59d4b61a6be3",
   "metadata": {},
   "source": [
    "## Output files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b5253102-d4b1-4fc7-9833-e4e58dabc299",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_id_to_vid_file = os.path.join(genie_dir, \"genie.vcf_id_to_hgvsg.tsv.gz\")\n",
    "annot_file = os.path.join(genie_dir, \"genie.annot.tsv.gz\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd4a3848-0b5a-46e5-87f7-f80df6b5f230",
   "metadata": {},
   "source": [
    "## Nirvana version and annotation data sources"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "47c2033c-40ab-41e9-b900-7b931c85879c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>GENIE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>annotator</td>\n",
       "      <td>Nirvana 3.21.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>creationTime</td>\n",
       "      <td>2024-03-15 13:55:27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>genomeAssembly</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>schemaVersion</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>dataVersion</td>\n",
       "      <td>73</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Attribute                GENIE\n",
       "0       annotator       Nirvana 3.21.0\n",
       "1    creationTime  2024-03-15 13:55:27\n",
       "2  genomeAssembly               GRCh37\n",
       "3   schemaVersion                    6\n",
       "4     dataVersion                   73"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nip.get_header_scalars(json_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c708dab1-6797-40d1-bbcf-a69be77e8ef3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>version</th>\n",
       "      <th>description</th>\n",
       "      <th>releaseDate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ensembl</td>\n",
       "      <td>110</td>\n",
       "      <td>Ensembl GRCh37 Release</td>\n",
       "      <td>2023-02-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RefSeq</td>\n",
       "      <td>105.20220307</td>\n",
       "      <td>NCBI Homo sapiens Annotation Release</td>\n",
       "      <td>2022-03-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ClinVar</td>\n",
       "      <td>20231028</td>\n",
       "      <td>A freely accessible, public archive of reports...</td>\n",
       "      <td>2023-11-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>COSMIC</td>\n",
       "      <td>98</td>\n",
       "      <td>resource for exploring the impact of somatic m...</td>\n",
       "      <td>2023-05-23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>dbSNP</td>\n",
       "      <td>156</td>\n",
       "      <td>Identifiers for observed variants</td>\n",
       "      <td>2023-09-28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>dbSNP</td>\n",
       "      <td>151</td>\n",
       "      <td>Identifiers for observed variants</td>\n",
       "      <td>2018-04-23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GME</td>\n",
       "      <td>20160618</td>\n",
       "      <td>Greater Middle Eastern Variome allele frequenc...</td>\n",
       "      <td>2016-06-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>gnomAD</td>\n",
       "      <td>2.1</td>\n",
       "      <td>Allele frequencies from Genome Aggregation Dat...</td>\n",
       "      <td>2018-10-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MITOMAP</td>\n",
       "      <td>20200819</td>\n",
       "      <td>Small variants in the MITOMAP human mitochondr...</td>\n",
       "      <td>2020-08-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1000 Genomes Project</td>\n",
       "      <td>Phase 3 v5a</td>\n",
       "      <td>A public catalogue of human variation and geno...</td>\n",
       "      <td>2013-05-27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>REVEL</td>\n",
       "      <td>20200205</td>\n",
       "      <td>Pathogenicity scores of missense variants pred...</td>\n",
       "      <td>2020-02-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>TOPMed</td>\n",
       "      <td>freeze_5</td>\n",
       "      <td>Allele frequencies from TOPMed data lifted ove...</td>\n",
       "      <td>2017-08-28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ClinGen</td>\n",
       "      <td>20160414</td>\n",
       "      <td>None</td>\n",
       "      <td>2016-04-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ClinGen Dosage Sensitivity Map</td>\n",
       "      <td>20231105</td>\n",
       "      <td>Dosage sensitivity map from ClinGen (dbVar)</td>\n",
       "      <td>2023-11-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>DECIPHER</td>\n",
       "      <td>201509</td>\n",
       "      <td>Common copy-number variants and their frequenc...</td>\n",
       "      <td>2015-09-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>gnomAD_SV</td>\n",
       "      <td>2.1</td>\n",
       "      <td>A structural variation reference from Genome A...</td>\n",
       "      <td>2020-08-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>MITOMAP_SV</td>\n",
       "      <td>20200819</td>\n",
       "      <td>Large structural variants in the MITOMAP human...</td>\n",
       "      <td>2020-08-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1000 Genomes Project (SV)</td>\n",
       "      <td>Phase 3 v5a</td>\n",
       "      <td>A public catalogue of human variation and geno...</td>\n",
       "      <td>2013-05-27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>COSMIC gene fusions</td>\n",
       "      <td>98</td>\n",
       "      <td>manually curated somatic gene fusions</td>\n",
       "      <td>2023-05-23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>FusionCatcher</td>\n",
       "      <td>1.33</td>\n",
       "      <td>known germline and somatic gene fusions</td>\n",
       "      <td>2020-12-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>DANN</td>\n",
       "      <td>20200205</td>\n",
       "      <td>Pathogenicity scores of missense variants pred...</td>\n",
       "      <td>2020-02-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Gerp</td>\n",
       "      <td>20110522</td>\n",
       "      <td>GERP identifies constrained elements in multip...</td>\n",
       "      <td>2011-05-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>gnomAD_gene_scores</td>\n",
       "      <td>2.1</td>\n",
       "      <td>Gene Loss of function scores from Genome Aggre...</td>\n",
       "      <td>2018-10-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>CosmicSv</td>\n",
       "      <td>98</td>\n",
       "      <td>Cosmic Structural Variants</td>\n",
       "      <td>2023-05-23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ClinGen disease validity curations</td>\n",
       "      <td>20231105</td>\n",
       "      <td>Disease validity curations from ClinGen (dbVar)</td>\n",
       "      <td>2023-11-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>phyloP</td>\n",
       "      <td>hg19</td>\n",
       "      <td>46 way conservation score between humans and 4...</td>\n",
       "      <td>2009-11-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>gnomAD_LCR</td>\n",
       "      <td>2.1</td>\n",
       "      <td>Low complexity regions from gnomAD.</td>\n",
       "      <td>2018-03-28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>MitochondrialHeteroplasmy</td>\n",
       "      <td>20180410</td>\n",
       "      <td>Variant read frequency percentiles for the Mit...</td>\n",
       "      <td>2020-05-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>CancerHotspots</td>\n",
       "      <td>2017</td>\n",
       "      <td>A resouce for statistically significant mutati...</td>\n",
       "      <td>2017-01-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  name       version  \\\n",
       "0                              Ensembl           110   \n",
       "1                               RefSeq  105.20220307   \n",
       "2                              ClinVar      20231028   \n",
       "3                               COSMIC            98   \n",
       "4                                dbSNP           156   \n",
       "5                                dbSNP           151   \n",
       "6                                  GME      20160618   \n",
       "7                               gnomAD           2.1   \n",
       "8                              MITOMAP      20200819   \n",
       "9                 1000 Genomes Project   Phase 3 v5a   \n",
       "10                               REVEL      20200205   \n",
       "11                              TOPMed      freeze_5   \n",
       "12                             ClinGen      20160414   \n",
       "13      ClinGen Dosage Sensitivity Map      20231105   \n",
       "14                            DECIPHER        201509   \n",
       "15                           gnomAD_SV           2.1   \n",
       "16                          MITOMAP_SV      20200819   \n",
       "17           1000 Genomes Project (SV)   Phase 3 v5a   \n",
       "18                 COSMIC gene fusions            98   \n",
       "19                       FusionCatcher          1.33   \n",
       "20                                DANN      20200205   \n",
       "21                                Gerp      20110522   \n",
       "22                  gnomAD_gene_scores           2.1   \n",
       "23                            CosmicSv            98   \n",
       "24  ClinGen disease validity curations      20231105   \n",
       "25                              phyloP          hg19   \n",
       "26                          gnomAD_LCR           2.1   \n",
       "27           MitochondrialHeteroplasmy      20180410   \n",
       "28                      CancerHotspots          2017   \n",
       "\n",
       "                                          description releaseDate  \n",
       "0                              Ensembl GRCh37 Release  2023-02-08  \n",
       "1                NCBI Homo sapiens Annotation Release  2022-03-10  \n",
       "2   A freely accessible, public archive of reports...  2023-11-05  \n",
       "3   resource for exploring the impact of somatic m...  2023-05-23  \n",
       "4                   Identifiers for observed variants  2023-09-28  \n",
       "5                   Identifiers for observed variants  2018-04-23  \n",
       "6   Greater Middle Eastern Variome allele frequenc...  2016-06-18  \n",
       "7   Allele frequencies from Genome Aggregation Dat...  2018-10-17  \n",
       "8   Small variants in the MITOMAP human mitochondr...  2020-08-19  \n",
       "9   A public catalogue of human variation and geno...  2013-05-27  \n",
       "10  Pathogenicity scores of missense variants pred...  2020-02-05  \n",
       "11  Allele frequencies from TOPMed data lifted ove...  2017-08-28  \n",
       "12                                               None  2016-04-14  \n",
       "13        Dosage sensitivity map from ClinGen (dbVar)  2023-11-05  \n",
       "14  Common copy-number variants and their frequenc...  2015-09-01  \n",
       "15  A structural variation reference from Genome A...  2020-08-20  \n",
       "16  Large structural variants in the MITOMAP human...  2020-08-19  \n",
       "17  A public catalogue of human variation and geno...  2013-05-27  \n",
       "18              manually curated somatic gene fusions  2023-05-23  \n",
       "19            known germline and somatic gene fusions  2020-12-22  \n",
       "20  Pathogenicity scores of missense variants pred...  2020-02-05  \n",
       "21  GERP identifies constrained elements in multip...  2011-05-22  \n",
       "22  Gene Loss of function scores from Genome Aggre...  2018-10-17  \n",
       "23                         Cosmic Structural Variants  2023-05-23  \n",
       "24    Disease validity curations from ClinGen (dbVar)  2023-11-05  \n",
       "25  46 way conservation score between humans and 4...  2009-11-10  \n",
       "26                Low complexity regions from gnomAD.  2018-03-28  \n",
       "27  Variant read frequency percentiles for the Mit...  2020-05-21  \n",
       "28  A resouce for statistically significant mutati...  2017-01-01  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nip.get_data_sources(json_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad2a7f8a-e838-4e9e-b1f8-08dc6d067b06",
   "metadata": {},
   "source": [
    "## Get a mapping from Niravana variant ids to VCF-IDs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bacd71fa-1a49-4e48-9736-c3917f47505a",
   "metadata": {},
   "source": [
    "### Function for reading a VCF file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b99e3078-f302-40a3-beac-8c1c0f56cec5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_vcf_gz(file_name):\n",
    "    with gzip.open(file_name, \"rt\") as f:\n",
    "        header = f.readline()\n",
    "        while not header.startswith(\"#CHROM\"):\n",
    "            header = f.readline()\n",
    "        header = header[1:].strip().split(\"\\t\")\n",
    "    df = pd.read_csv(file_name, comment=\"#\", header=None, sep=\"\\t\", compression=\"gzip\")\n",
    "    df.columns = header\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1099010-0b44-4c13-aea2-ade361bac0a6",
   "metadata": {},
   "source": [
    "### Read the vcf file "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "01c8494b-91f5-4986-9f00-3db5b1157d5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf = read_vcf_gz(vcf_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b5472df3-1605-4744-b5f2-33a69b82a67b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>GENIE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1684345</td>\n",
       "      <td>c431afab69588f3150a378d08fc0bf8713584d5eb092b9...</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:VF</td>\n",
       "      <td>1/1:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1684347</td>\n",
       "      <td>adc1644ed30da5964fd03a592c73e5ddf0d78e924e62d4...</td>\n",
       "      <td>C</td>\n",
       "      <td>CCCTCCT</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:VF</td>\n",
       "      <td>1/1:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1684347</td>\n",
       "      <td>a86d0a0f81cff8daa9f32b83497a784463aab899b34f84...</td>\n",
       "      <td>CCCTCCT</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:VF</td>\n",
       "      <td>1/1:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1684391</td>\n",
       "      <td>97b2d690a7032356e8d65641b60d3fb891fd1770e1c1b6...</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:VF</td>\n",
       "      <td>1/1:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1684395</td>\n",
       "      <td>15ae11efa9729592af42a0f345ee339da9979e129d44a7...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:VF</td>\n",
       "      <td>1/1:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862092</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21894633</td>\n",
       "      <td>4d2d45c22fb442113473f832cecc3e47544ad8a91362c3...</td>\n",
       "      <td>CAA</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:VF</td>\n",
       "      <td>1/1:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862093</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21894633</td>\n",
       "      <td>992feda83adb1109d0569dfb56aa523112d5abf4f80cf3...</td>\n",
       "      <td>CAAA</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:VF</td>\n",
       "      <td>1/1:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862094</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21897244</td>\n",
       "      <td>dff14a4f48c0a693839b97c747dc329e97d32cc2fc537f...</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:VF</td>\n",
       "      <td>1/1:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862095</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21901521</td>\n",
       "      <td>4529089ce4ace971769eaff8da439021a06c49d903692a...</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:VF</td>\n",
       "      <td>1/1:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862096</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21903282</td>\n",
       "      <td>e64a1b69180d7ecf53734608eef6129608499ded19dd11...</td>\n",
       "      <td>GT</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:VF</td>\n",
       "      <td>1/1:1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>862097 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       CHROM       POS                                                 ID  \\\n",
       "0       chr1   1684345  c431afab69588f3150a378d08fc0bf8713584d5eb092b9...   \n",
       "1       chr1   1684347  adc1644ed30da5964fd03a592c73e5ddf0d78e924e62d4...   \n",
       "2       chr1   1684347  a86d0a0f81cff8daa9f32b83497a784463aab899b34f84...   \n",
       "3       chr1   1684391  97b2d690a7032356e8d65641b60d3fb891fd1770e1c1b6...   \n",
       "4       chr1   1684395  15ae11efa9729592af42a0f345ee339da9979e129d44a7...   \n",
       "...      ...       ...                                                ...   \n",
       "862092  chrY  21894633  4d2d45c22fb442113473f832cecc3e47544ad8a91362c3...   \n",
       "862093  chrY  21894633  992feda83adb1109d0569dfb56aa523112d5abf4f80cf3...   \n",
       "862094  chrY  21897244  dff14a4f48c0a693839b97c747dc329e97d32cc2fc537f...   \n",
       "862095  chrY  21901521  4529089ce4ace971769eaff8da439021a06c49d903692a...   \n",
       "862096  chrY  21903282  e64a1b69180d7ecf53734608eef6129608499ded19dd11...   \n",
       "\n",
       "            REF      ALT QUAL FILTER INFO FORMAT  GENIE  \n",
       "0             A        T    .   PASS    .  GT:VF  1/1:1  \n",
       "1             C  CCCTCCT    .   PASS    .  GT:VF  1/1:1  \n",
       "2       CCCTCCT        C    .   PASS    .  GT:VF  1/1:1  \n",
       "3             C        A    .   PASS    .  GT:VF  1/1:1  \n",
       "4             C        T    .   PASS    .  GT:VF  1/1:1  \n",
       "...         ...      ...  ...    ...  ...    ...    ...  \n",
       "862092      CAA        C    .   PASS    .  GT:VF  1/1:1  \n",
       "862093     CAAA        C    .   PASS    .  GT:VF  1/1:1  \n",
       "862094        G        T    .   PASS    .  GT:VF  1/1:1  \n",
       "862095        C        G    .   PASS    .  GT:VF  1/1:1  \n",
       "862096       GT        G    .   PASS    .  GT:VF  1/1:1  \n",
       "\n",
       "[862097 rows x 10 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c515315d-edc3-45bc-be15-735a6d319fc7",
   "metadata": {},
   "source": [
    "### Get a mapping of CHROM, POS, REF, ALT to Nirvana variant ID"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b71f540f-7f4a-4d5c-8b54-ebba2119a1f6",
   "metadata": {},
   "source": [
    "All positions from Nirvana's output are scanned and a mapping table from the original specification of chromosome, position, reference and alternative allele (from input VCF file) to Nirvana's variant identifier is created."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "440432cf-9bd6-4824-a047-941f1c3202df",
   "metadata": {},
   "outputs": [],
   "source": [
    "pos_all = nip.get_positions(json_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "be8c49d8-d68d-4531-b6b4-1c8f2a7b5da3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "aebfdc76c9934790a8f63fbc5e061784",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/862097 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "rows = []\n",
    "for p in tqdm(pos_all):\n",
    "    if len(p[\"altAlleles\"]) != 1:\n",
    "        raise ValueError(\"Not exactly one refAllele per position\")\n",
    "    if len(p[\"variants\"]) != 1:\n",
    "        raise ValueError(\"Not exactly one variant per position\")\n",
    "    rows.append(\n",
    "        {\n",
    "            \"CHROM\": p[\"chromosome\"],\n",
    "            \"POS\": p[\"position\"],\n",
    "            \"REF\": p[\"refAllele\"],\n",
    "            \"ALT\": p[\"altAlleles\"][0],\n",
    "            \"vid\": p[\"variants\"][0][\"vid\"],\n",
    "            \"hgvsg\": p[\"variants\"][0].get(\"hgvsg\", pd.NA),\n",
    "            \"sample\": p[\"samples\"][0][\"sampleId\"],\n",
    "        }\n",
    "    )\n",
    "cpra_to_vid = pd.DataFrame(rows)\n",
    "del rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e01a5b48-1648-4403-959d-2583931ce547",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>vid</th>\n",
       "      <th>hgvsg</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1684345</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>1-1684345-A-T</td>\n",
       "      <td>NC_000001.10:g.1684345A&gt;T</td>\n",
       "      <td>GENIE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1684347</td>\n",
       "      <td>C</td>\n",
       "      <td>CCCTCCT</td>\n",
       "      <td>1-1684347-C-CCCTCCT</td>\n",
       "      <td>NC_000001.10:g.1684370_1684375dup</td>\n",
       "      <td>GENIE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1684347</td>\n",
       "      <td>CCCTCCT</td>\n",
       "      <td>C</td>\n",
       "      <td>1-1684347-CCCTCCT-C</td>\n",
       "      <td>NC_000001.10:g.1684370_1684375del</td>\n",
       "      <td>GENIE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1684391</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1-1684391-C-A</td>\n",
       "      <td>NC_000001.10:g.1684391C&gt;A</td>\n",
       "      <td>GENIE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1684395</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1-1684395-C-T</td>\n",
       "      <td>NC_000001.10:g.1684395C&gt;T</td>\n",
       "      <td>GENIE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862092</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21894633</td>\n",
       "      <td>CAA</td>\n",
       "      <td>C</td>\n",
       "      <td>Y-21894633-CAA-C</td>\n",
       "      <td>NC_000024.9:g.21894647_21894648del</td>\n",
       "      <td>GENIE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862093</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21894633</td>\n",
       "      <td>CAAA</td>\n",
       "      <td>C</td>\n",
       "      <td>Y-21894633-CAAA-C</td>\n",
       "      <td>NC_000024.9:g.21894646_21894648del</td>\n",
       "      <td>GENIE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862094</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21897244</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Y-21897244-G-T</td>\n",
       "      <td>NC_000024.9:g.21897244G&gt;T</td>\n",
       "      <td>GENIE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862095</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21901521</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Y-21901521-C-G</td>\n",
       "      <td>NC_000024.9:g.21901521C&gt;G</td>\n",
       "      <td>GENIE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862096</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21903282</td>\n",
       "      <td>GT</td>\n",
       "      <td>G</td>\n",
       "      <td>Y-21903282-GT-G</td>\n",
       "      <td>NC_000024.9:g.21903287del</td>\n",
       "      <td>GENIE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>862097 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       CHROM       POS      REF      ALT                  vid  \\\n",
       "0       chr1   1684345        A        T        1-1684345-A-T   \n",
       "1       chr1   1684347        C  CCCTCCT  1-1684347-C-CCCTCCT   \n",
       "2       chr1   1684347  CCCTCCT        C  1-1684347-CCCTCCT-C   \n",
       "3       chr1   1684391        C        A        1-1684391-C-A   \n",
       "4       chr1   1684395        C        T        1-1684395-C-T   \n",
       "...      ...       ...      ...      ...                  ...   \n",
       "862092  chrY  21894633      CAA        C     Y-21894633-CAA-C   \n",
       "862093  chrY  21894633     CAAA        C    Y-21894633-CAAA-C   \n",
       "862094  chrY  21897244        G        T       Y-21897244-G-T   \n",
       "862095  chrY  21901521        C        G       Y-21901521-C-G   \n",
       "862096  chrY  21903282       GT        G      Y-21903282-GT-G   \n",
       "\n",
       "                                     hgvsg sample  \n",
       "0                NC_000001.10:g.1684345A>T  GENIE  \n",
       "1        NC_000001.10:g.1684370_1684375dup  GENIE  \n",
       "2        NC_000001.10:g.1684370_1684375del  GENIE  \n",
       "3                NC_000001.10:g.1684391C>A  GENIE  \n",
       "4                NC_000001.10:g.1684395C>T  GENIE  \n",
       "...                                    ...    ...  \n",
       "862092  NC_000024.9:g.21894647_21894648del  GENIE  \n",
       "862093  NC_000024.9:g.21894646_21894648del  GENIE  \n",
       "862094           NC_000024.9:g.21897244G>T  GENIE  \n",
       "862095           NC_000024.9:g.21901521C>G  GENIE  \n",
       "862096           NC_000024.9:g.21903287del  GENIE  \n",
       "\n",
       "[862097 rows x 7 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cpra_to_vid"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96d8624f-32ce-4ce8-ae6d-a868f00f69c5",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Create a mapping table of VCF id to hgvsg specification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "91b062a7-29a0-444f-a598-713663e8ad77",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>sample</th>\n",
       "      <th>vid</th>\n",
       "      <th>hgvsg</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>c431afab69588f3150a378d08fc0bf8713584d5eb092b9...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1684345</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>GENIE</td>\n",
       "      <td>1-1684345-A-T</td>\n",
       "      <td>NC_000001.10:g.1684345A&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>adc1644ed30da5964fd03a592c73e5ddf0d78e924e62d4...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1684347</td>\n",
       "      <td>C</td>\n",
       "      <td>CCCTCCT</td>\n",
       "      <td>GENIE</td>\n",
       "      <td>1-1684347-C-CCCTCCT</td>\n",
       "      <td>NC_000001.10:g.1684370_1684375dup</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>a86d0a0f81cff8daa9f32b83497a784463aab899b34f84...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1684347</td>\n",
       "      <td>CCCTCCT</td>\n",
       "      <td>C</td>\n",
       "      <td>GENIE</td>\n",
       "      <td>1-1684347-CCCTCCT-C</td>\n",
       "      <td>NC_000001.10:g.1684370_1684375del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>97b2d690a7032356e8d65641b60d3fb891fd1770e1c1b6...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1684391</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE</td>\n",
       "      <td>1-1684391-C-A</td>\n",
       "      <td>NC_000001.10:g.1684391C&gt;A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>15ae11efa9729592af42a0f345ee339da9979e129d44a7...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1684395</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GENIE</td>\n",
       "      <td>1-1684395-C-T</td>\n",
       "      <td>NC_000001.10:g.1684395C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>864954</th>\n",
       "      <td>4d2d45c22fb442113473f832cecc3e47544ad8a91362c3...</td>\n",
       "      <td>chrY</td>\n",
       "      <td>21894633</td>\n",
       "      <td>CAA</td>\n",
       "      <td>C</td>\n",
       "      <td>GENIE</td>\n",
       "      <td>Y-21894633-CAA-C</td>\n",
       "      <td>NC_000024.9:g.21894647_21894648del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>864955</th>\n",
       "      <td>992feda83adb1109d0569dfb56aa523112d5abf4f80cf3...</td>\n",
       "      <td>chrY</td>\n",
       "      <td>21894633</td>\n",
       "      <td>CAAA</td>\n",
       "      <td>C</td>\n",
       "      <td>GENIE</td>\n",
       "      <td>Y-21894633-CAAA-C</td>\n",
       "      <td>NC_000024.9:g.21894646_21894648del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>864956</th>\n",
       "      <td>dff14a4f48c0a693839b97c747dc329e97d32cc2fc537f...</td>\n",
       "      <td>chrY</td>\n",
       "      <td>21897244</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>GENIE</td>\n",
       "      <td>Y-21897244-G-T</td>\n",
       "      <td>NC_000024.9:g.21897244G&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>864957</th>\n",
       "      <td>4529089ce4ace971769eaff8da439021a06c49d903692a...</td>\n",
       "      <td>chrY</td>\n",
       "      <td>21901521</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>GENIE</td>\n",
       "      <td>Y-21901521-C-G</td>\n",
       "      <td>NC_000024.9:g.21901521C&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>864958</th>\n",
       "      <td>e64a1b69180d7ecf53734608eef6129608499ded19dd11...</td>\n",
       "      <td>chrY</td>\n",
       "      <td>21903282</td>\n",
       "      <td>GT</td>\n",
       "      <td>G</td>\n",
       "      <td>GENIE</td>\n",
       "      <td>Y-21903282-GT-G</td>\n",
       "      <td>NC_000024.9:g.21903287del</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>862097 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                       ID CHROM       POS  \\\n",
       "0       c431afab69588f3150a378d08fc0bf8713584d5eb092b9...  chr1   1684345   \n",
       "1       adc1644ed30da5964fd03a592c73e5ddf0d78e924e62d4...  chr1   1684347   \n",
       "2       a86d0a0f81cff8daa9f32b83497a784463aab899b34f84...  chr1   1684347   \n",
       "3       97b2d690a7032356e8d65641b60d3fb891fd1770e1c1b6...  chr1   1684391   \n",
       "4       15ae11efa9729592af42a0f345ee339da9979e129d44a7...  chr1   1684395   \n",
       "...                                                   ...   ...       ...   \n",
       "864954  4d2d45c22fb442113473f832cecc3e47544ad8a91362c3...  chrY  21894633   \n",
       "864955  992feda83adb1109d0569dfb56aa523112d5abf4f80cf3...  chrY  21894633   \n",
       "864956  dff14a4f48c0a693839b97c747dc329e97d32cc2fc537f...  chrY  21897244   \n",
       "864957  4529089ce4ace971769eaff8da439021a06c49d903692a...  chrY  21901521   \n",
       "864958  e64a1b69180d7ecf53734608eef6129608499ded19dd11...  chrY  21903282   \n",
       "\n",
       "            REF      ALT sample                  vid  \\\n",
       "0             A        T  GENIE        1-1684345-A-T   \n",
       "1             C  CCCTCCT  GENIE  1-1684347-C-CCCTCCT   \n",
       "2       CCCTCCT        C  GENIE  1-1684347-CCCTCCT-C   \n",
       "3             C        A  GENIE        1-1684391-C-A   \n",
       "4             C        T  GENIE        1-1684395-C-T   \n",
       "...         ...      ...    ...                  ...   \n",
       "864954      CAA        C  GENIE     Y-21894633-CAA-C   \n",
       "864955     CAAA        C  GENIE    Y-21894633-CAAA-C   \n",
       "864956        G        T  GENIE       Y-21897244-G-T   \n",
       "864957        C        G  GENIE       Y-21901521-C-G   \n",
       "864958       GT        G  GENIE      Y-21903282-GT-G   \n",
       "\n",
       "                                     hgvsg  \n",
       "0                NC_000001.10:g.1684345A>T  \n",
       "1        NC_000001.10:g.1684370_1684375dup  \n",
       "2        NC_000001.10:g.1684370_1684375del  \n",
       "3                NC_000001.10:g.1684391C>A  \n",
       "4                NC_000001.10:g.1684395C>T  \n",
       "...                                    ...  \n",
       "864954  NC_000024.9:g.21894647_21894648del  \n",
       "864955  NC_000024.9:g.21894646_21894648del  \n",
       "864956           NC_000024.9:g.21897244G>T  \n",
       "864957           NC_000024.9:g.21901521C>G  \n",
       "864958           NC_000024.9:g.21903287del  \n",
       "\n",
       "[862097 rows x 8 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = vcf.copy().drop(columns=[\"QUAL\", \"FILTER\", \"INFO\", \"FORMAT\"])\n",
    "df = df.melt(\n",
    "    id_vars=[\"ID\", \"CHROM\", \"POS\", \"REF\", \"ALT\"], var_name=\"sample\", value_name=\"mut\"\n",
    ")\n",
    "df[\"mut\"] = df[\"mut\"].str.startswith(\"1/1\")\n",
    "df = df.loc[df.mut].drop(columns=\"mut\").reset_index(drop=True)\n",
    "id_to_hgvsg = df.merge(\n",
    "    cpra_to_vid, how=\"left\", on=[\"CHROM\", \"POS\", \"REF\", \"ALT\", \"sample\"]\n",
    ").drop_duplicates()\n",
    "id_to_hgvsg"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e19b00a-039a-4a7f-b699-f9b1b1c7319e",
   "metadata": {},
   "source": [
    "Assert that all variant IDs from the VCF file have a mapping to a HGVSG:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "127f42d1-0e46-4f4f-8b63-61ae1a30d2af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(vcf.ID) - set(id_to_hgvsg.ID))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d594c995-444d-4d9f-bbfc-8ceaa9b00fba",
   "metadata": {},
   "source": [
    "### Write the mapping table of VCF id to hgvsg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "491fe243-12ff-45f8-a93e-6b1c9a1641b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "id_to_hgvsg.to_csv(vcf_id_to_vid_file, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de87f576-8f94-40ef-a270-87d478ee735b",
   "metadata": {},
   "source": [
    "## Read the Cosmic Cancer Mutation Census file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "27c1ee0b-06f5-45c2-b38d-2a524d668e67",
   "metadata": {},
   "outputs": [],
   "source": [
    "def tier_rank(x):\n",
    "    tier_ranks = {\"1\": 1, \"2\": 2, \"3\": 3, \"Other\": 4, pd.NA: 5}\n",
    "    return tier_ranks[x]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0c897fab-e6bf-4134-b833-50b3ef088f75",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>GENE_NAME</th>\n",
       "      <th>ONC_TSG</th>\n",
       "      <th>CGC_TIER</th>\n",
       "      <th>MUTATION_SIGNIFICANCE_TIER</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENOMIC_MUTATION_ID</th>\n",
       "      <th>Mutation CDS</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>COSV100000072</th>\n",
       "      <th>c.206T&gt;C</th>\n",
       "      <td>MYOD1</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>1</td>\n",
       "      <td>Other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COSV100000124</th>\n",
       "      <th>c.177C&gt;A</th>\n",
       "      <td>MYOD1</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>1</td>\n",
       "      <td>Other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COSV100000125</th>\n",
       "      <th>c.355C&gt;T</th>\n",
       "      <td>MYOD1</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>1</td>\n",
       "      <td>Other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COSV100000126</th>\n",
       "      <th>c.296A&gt;G</th>\n",
       "      <td>MYOD1</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>1</td>\n",
       "      <td>Other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COSV100000133</th>\n",
       "      <th>c.790C&gt;A</th>\n",
       "      <td>MYOD1</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>1</td>\n",
       "      <td>Other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COSV99994110</th>\n",
       "      <th>c.122C&gt;A</th>\n",
       "      <td>CEP89</td>\n",
       "      <td>fusion</td>\n",
       "      <td>2</td>\n",
       "      <td>Other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COSV99994134</th>\n",
       "      <th>c.64C&gt;T</th>\n",
       "      <td>CEP89</td>\n",
       "      <td>fusion</td>\n",
       "      <td>2</td>\n",
       "      <td>Other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COSV99994197</th>\n",
       "      <th>c.889G&gt;A</th>\n",
       "      <td>CEP89</td>\n",
       "      <td>fusion</td>\n",
       "      <td>2</td>\n",
       "      <td>Other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COSV99994203</th>\n",
       "      <th>c.214C&gt;T</th>\n",
       "      <td>CEP89</td>\n",
       "      <td>fusion</td>\n",
       "      <td>2</td>\n",
       "      <td>Other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COSV99994246</th>\n",
       "      <th>c.1164G&gt;T</th>\n",
       "      <td>CEP89</td>\n",
       "      <td>fusion</td>\n",
       "      <td>2</td>\n",
       "      <td>Other</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>402968 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 GENE_NAME   ONC_TSG CGC_TIER  \\\n",
       "GENOMIC_MUTATION_ID Mutation CDS                                \n",
       "COSV100000072       c.206T>C         MYOD1  oncogene        1   \n",
       "COSV100000124       c.177C>A         MYOD1  oncogene        1   \n",
       "COSV100000125       c.355C>T         MYOD1  oncogene        1   \n",
       "COSV100000126       c.296A>G         MYOD1  oncogene        1   \n",
       "COSV100000133       c.790C>A         MYOD1  oncogene        1   \n",
       "...                                    ...       ...      ...   \n",
       "COSV99994110        c.122C>A         CEP89    fusion        2   \n",
       "COSV99994134        c.64C>T          CEP89    fusion        2   \n",
       "COSV99994197        c.889G>A         CEP89    fusion        2   \n",
       "COSV99994203        c.214C>T         CEP89    fusion        2   \n",
       "COSV99994246        c.1164G>T        CEP89    fusion        2   \n",
       "\n",
       "                                 MUTATION_SIGNIFICANCE_TIER  \n",
       "GENOMIC_MUTATION_ID Mutation CDS                             \n",
       "COSV100000072       c.206T>C                          Other  \n",
       "COSV100000124       c.177C>A                          Other  \n",
       "COSV100000125       c.355C>T                          Other  \n",
       "COSV100000126       c.296A>G                          Other  \n",
       "COSV100000133       c.790C>A                          Other  \n",
       "...                                                     ...  \n",
       "COSV99994110        c.122C>A                          Other  \n",
       "COSV99994134        c.64C>T                           Other  \n",
       "COSV99994197        c.889G>A                          Other  \n",
       "COSV99994203        c.214C>T                          Other  \n",
       "COSV99994246        c.1164G>T                         Other  \n",
       "\n",
       "[402968 rows x 4 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cmc = pd.read_table(cosmic_cmc_file, encoding=\"latin1\", low_memory=False)\n",
    "cmc = cmc[cmc.GENOMIC_MUTATION_ID.notna()]\n",
    "cmc = cmc[\n",
    "    [\"GENOMIC_MUTATION_ID\", \"GENE_NAME\", \"Mutation CDS\", \"ONC_TSG\", \"CGC_TIER\", \"MUTATION_SIGNIFICANCE_TIER\"]\n",
    "].drop_duplicates()\n",
    "cmc = cmc.groupby([\"GENOMIC_MUTATION_ID\", \"GENE_NAME\", \"Mutation CDS\", \"ONC_TSG\", \"CGC_TIER\"])[\n",
    "    \"MUTATION_SIGNIFICANCE_TIER\"\n",
    "].apply(lambda x: min(x, key=tier_rank))\n",
    "cmc = cmc.reset_index().set_index([\"GENOMIC_MUTATION_ID\", \"Mutation CDS\"], drop=True).sort_index()\n",
    "cmc[\"CGC_TIER\"] = cmc[\"CGC_TIER\"].astype(\"int\").astype(\"str\")\n",
    "cmc"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01ba5ca2-0903-47fa-8f65-75967ea02f7c",
   "metadata": {},
   "source": [
    "## Get annotated table of individual mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "05f24a99-f0e6-4f10-9796-209084ea59dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "pos = nip.filter_positions_by_variants(pos_all, lambda v: \"transcripts\" in v)\n",
    "# pos = nip.filter_positions_by_variants(pos, lambda v: nip.common_variant_filter(v, 5e-4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "9a07bd85-a4b5-416a-9455-8f651601554a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "862097"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(pos_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "9052d295-7936-4d36-9bc9-ef2ed861218f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "861517"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(pos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4f16b528-00be-498a-8cad-eab861fa1467",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d8a94bcf8a504d65bac2736cb4a16da0",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Positions:   0%|          | 0/861517 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mt = nip.get_mutation_table_for_positions(pos)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94b2c0e9-adf5-4bbb-8d99-1a9ba907dd20",
   "metadata": {},
   "source": [
    "### Add Cosmic Cancer Gene and Mutation Census Tiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "db837b41-080d-4ffc-9a5c-8ac10045ad79",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6a91668bfeb3447fa48759566b247afc",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/861517 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mutation_tier_ranks = {\"1\": 1, \"2\": 2, \"3\": 3, \"Other\": 4, pd.NA: 5}\n",
    "gene_tier_ranks = {\"1\": 1, \"2\": 2, pd.NA: 3}\n",
    "cmc_tiers = []\n",
    "for p in tqdm(pos):\n",
    "    for v in p[\"variants\"]:\n",
    "        vid = v[\"vid\"]\n",
    "        for t in v.get(\"transcripts\", []):\n",
    "            best_mutation_tier = pd.NA\n",
    "            best_gene_type = pd.NA\n",
    "            best_gene_tier = pd.NA\n",
    "            best_gene_name = pd.NA\n",
    "            # Only investigate transcripts with a cDNA change\n",
    "            if \"hgvsc\" not in t:\n",
    "                continue\n",
    "            hgvsc_short = t[\"hgvsc\"].split(\":\")[1]\n",
    "            for cosmic_entry in v.get(\"cosmic\", []):\n",
    "                # Only keep COSMIC entries for the same genomic change\n",
    "                if not cosmic_entry.get(\"isAlleleSpecific\", False):\n",
    "                    continue\n",
    "                # Only keep COSMIC entries which were found in the Cancer Mutation Census\n",
    "                cosmic_id = cosmic_entry[\"id\"]\n",
    "                if cosmic_id not in cmc.index:\n",
    "                    continue\n",
    "                cmc_entries = cmc.loc[cosmic_id]\n",
    "                #if hgvsc_short not in cmc_entries.index:\n",
    "                #    continue\n",
    "                #cmc_entries = cmc_entries.loc[[hgvsc_short]]\n",
    "                for idx, cmc_entry in cmc_entries.iterrows():\n",
    "                    cmc_tier = cmc_entry.MUTATION_SIGNIFICANCE_TIER\n",
    "                    cgc_gene_type = cmc_entry.ONC_TSG\n",
    "                    cgc_tier = cmc_entry.CGC_TIER\n",
    "                    cgc_gene = cmc_entry.GENE_NAME\n",
    "                    if mutation_tier_ranks[cmc_tier] < mutation_tier_ranks[best_mutation_tier]:\n",
    "                        best_mutation_tier = cmc_tier\n",
    "                    if gene_tier_ranks[cgc_tier] < gene_tier_ranks[best_gene_tier]:\n",
    "                        best_gene_type = cgc_gene_type\n",
    "                        best_gene_tier = cgc_tier\n",
    "                        best_gene_name = cgc_gene\n",
    "            cmc_tiers.append({\n",
    "                \"vid\": vid,\n",
    "                \"hgvsc\": t[\"hgvsc\"],\n",
    "                \"cosmicGeneName\": best_gene_name,\n",
    "                \"cosmicGeneType\": best_gene_type,\n",
    "                \"cosmicGeneTier\": best_gene_tier,\n",
    "                \"cosmicMutationTier\": best_mutation_tier\n",
    "            })\n",
    "cmc_tiers = pd.DataFrame(cmc_tiers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "bb1d9d24-5264-4243-a24b-d4ba0d2a16ca",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "mt = mt.merge(cmc_tiers, on=[\"vid\", \"hgvsc\"], how=\"left\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "3e174ef4-ebc4-4a0c-bba2-924010ca6e87",
   "metadata": {},
   "outputs": [],
   "source": [
    "mt.loc[\n",
    "    (mt.hgvsp.isna() | (mt.hgvsp == \"\")) & ~((mt.hgnc == mt.cosmicGeneName) & (mt.cosmicMutationTier.notna())),\n",
    "    [\"cosmicGeneName\", \"cosmicGeneType\", \"cosmicGeneTier\", \"cosmicMutationTier\"],\n",
    "] = pd.NA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ea15dc3-5faa-472a-8d80-f37860d8a0b3",
   "metadata": {},
   "source": [
    "So far a cancer gene cencus tier is only added if a variant has an assigned Cosmic entry. As a consequence, some cancer genes do not get a tier assigned if the variants found in the samples are not included in the cancer mutation census. Here we add an the tier based on gene names for those variants that don't have a tier assigned yet."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e6485464-e422-4b21-a272-1f22361e1c61",
   "metadata": {},
   "outputs": [],
   "source": [
    "cmc_gene_to_tier = (\n",
    "    cmc[[\"GENE_NAME\", \"ONC_TSG\", \"CGC_TIER\"]].drop_duplicates().set_index(\"GENE_NAME\", drop=True)\n",
    ")\n",
    "mt = mt.merge(cmc_gene_to_tier, how=\"left\", left_on=\"hgnc\", right_index=True)\n",
    "subset = mt.cosmicGeneTier.isna()\n",
    "mt.loc[subset, \"cosmicGeneType\"] = mt.loc[subset, \"ONC_TSG\"]\n",
    "mt.loc[subset, \"cosmicGeneTier\"] = mt.loc[subset, \"CGC_TIER\"]\n",
    "mt = mt.drop(columns=[\"ONC_TSG\", \"CGC_TIER\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7282c8d-a042-4061-9d7f-84e698f71032",
   "metadata": {},
   "source": [
    "### Add VEP consequence priority"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "2590ff87-5465-4d48-9580-fdbcd5a92b18",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_highest_vep_priority(consequences):\n",
    "    if type(consequences) is str:\n",
    "        consequences = consequences.split(\",\")\n",
    "    hp = min([nip.get_vep_priority_for_consequence(c) for c in consequences])\n",
    "    return hp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "eb95710c-efc7-47c8-80c1-1645bcf3c66e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "mt[\"consequence_priority\"] = mt.consequence.map(\n",
    "    lambda x: get_highest_vep_priority(x) if type(x) is list else x\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51062d2f-382a-4c68-a8f1-c9940f7976e9",
   "metadata": {},
   "source": [
    "### Write mutation table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "524066d1-059c-4e0d-929c-5b1800b382b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "mt[\"consequence\"] = mt[\"consequence\"].astype(\"str\")\n",
    "mt = mt.drop_duplicates()\n",
    "mt.to_csv(annot_file, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d88e2851-2de4-4d63-9623-3083f4aaad7c",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Timestamp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "18c42aad-94ab-4f4c-8722-9a29986a8714",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'2024-03-21 18:57:21 CET'"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from datetime import datetime\n",
    "\n",
    "datetime.now().astimezone().strftime(\"%Y-%m-%d %H:%M:%S %Z\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "37de3e59-715c-40bb-84d9-ce930c7ddbb2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----\n",
      "nirvanaparser       0.2.129\n",
      "pandas              2.1.4\n",
      "sinfo               0.3.1\n",
      "tqdm                4.66.1\n",
      "-----\n",
      "IPython             8.18.1\n",
      "jupyter_client      8.6.0\n",
      "jupyter_core        5.7.1\n",
      "jupyterlab          4.0.10\n",
      "notebook            7.0.6\n",
      "-----\n",
      "Python 3.12.0 | packaged by conda-forge | (main, Oct  3 2023, 08:43:22) [GCC 12.3.0]\n",
      "Linux-5.13.0-1029-aws-x86_64-with-glibc2.31\n",
      "8 logical CPU cores, x86_64\n",
      "-----\n",
      "Session information updated at 2024-03-21 18:57\n"
     ]
    }
   ],
   "source": [
    "from sinfo import sinfo\n",
    "sinfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "fdfa453b-1e21-4699-afbb-d8455d0ae558",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['sample', 'chromosome', 'position', 'genotype', 'variantFrequency',\n",
       "       'hgnc', 'source', 'geneId', 'transcriptId', 'proteinId', 'hgvsc',\n",
       "       'hgvsp', 'isCanonical', 'vid', 'hgvsg', 'begin', 'end', 'refAllele',\n",
       "       'altAllele', 'variantType', 'bioType', 'geneType', 'mutationStatus',\n",
       "       'codons', 'aminoAcids', 'cdnaPos', 'cdsPos', 'exons', 'proteinPos',\n",
       "       'consequence', 'cosmicSampleCount', 'maxGnomadAf', 'maxGnomadExomeAf',\n",
       "       'maxOneKG', 'cosmicGeneName', 'cosmicGeneType', 'cosmicGeneTier',\n",
       "       'cosmicMutationTier', 'consequence_priority'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mt.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22533c5a-3605-45eb-bf8a-3e367f1d854f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (nirvana)",
   "language": "python",
   "name": "nirvana"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
